### PR TITLE
Use DGP to connect logical switches to the cluster router.

### PR DESCRIPTION
**Pull request by hzhou8**
This patch sets gateway-chassis to the node itself for each logical
router port that connects node level switch to the cluster router. This
makes use of the OVN optimization [0] that avoids flooding-filling
unrelated datapaths by ovn-controller on each node, which perfectly
fits the ovn-kubernetes deployments where each logical switch is bound
to a chassis, significantly improving scalability of ovn-controller (10
times faster for recompute and 70-80% memory savings for ovn-controller
\+ OVS combined. See more details in [0]).

[0] https://github.com/ovn-org/ovn/commit/22298fd3790825f45680508bde821bb9a2d525d2

Signed-off-by: Han Zhou <hzhou@ovn.org>

### CODE REVIEW

File: go-controller/pkg/ovn/ovn.go
Comment (dcbw) [inline go-controller/pkg/ovn/ovn.go:908]: Shouldn't this get moved under the `} else if hostSubnets != nil {` just below?

Comment (dcbw) [reply go-controller/pkg/ovn/ovn.go:908]: Also, if we're going to use the chassisID outside of the gateway code we should just decouple that from the L3Gateway code on the node side. It's a separate annotation already, but handled by the gateway code. Should split that out.

Comment (hzhou8) [reply go-controller/pkg/ovn/ovn.go:908]: > Shouldn't this get moved under the `} else if hostSubnets != nil {` just below?

No, because this is regardless of the config.GatewayModeDisabled.

Comment (hzhou8) [reply go-controller/pkg/ovn/ovn.go:908]: > Also, if we're going to use the chassisID outside of the gateway code we should just decouple that from the L3Gateway code on the node side. It's a separate annotation already, but handled by the gateway code. Should split that out.

Agree. I am still getting familiar with the code, so I was trying to make as little change as possible. Let me see if I can change it that way with confidence.

**Comment by numansiddique** [comment]
OVN patche are merged - FYI

**Comment by hzhou8** [comment]
The CI failure is:
=== RUN   TestAdd/3:EgressFirewall_DNS_Run_Runs_update_after_the_ttl_returned_from_the_DNS_server_expires
    egressfirewall_dns_test.go:243: &{{0 0} 0xc00023bcc0 map[] 0xc000f13270 0xc001846f60 0xc001846fc0 0xc001846f00} <nil>
    egressfirewall_dns_test.go:275: timeout waiting for update based on ttl to fire or process

This timeout has nothing to do with the change, and it passes on my local environment. I suspect it is the test case problem. Shall we retry the CI job? @girishmg

**Comment by trozet** [comment]
@numansiddique @hzhou8 can we get a new koji build with the fixes and not use .dev dockerfile please?

**Comment by girishmg** [comment]
@trozet .dev Dockerfile change was to get a quick regression run. This MR cannot be merged until Han's OVN changes lands on Fedora

**Comment by hzhou8** [comment]
@trozet what do you mean by a new koji build? Where to put it?
@girishmg do you mean ovn-k8s upstream can only use OVN features from stable releases? I understand that ovn-k8s releases should depend on released OVN only, but is that required for upstream, too? I think that could potentially block new feature development for ovn-k8s. E.g. if there are multiple changes depending on OVN upstream changes, then all these changes may be blocked for a OVN release cycle.

**Comment by numansiddique** [comment]
I do agree too that upstream ovn-k8s should use upstream OVN.

Here is the fedora update which is synced with the latest OVN master.  Ideally I'd have preferred to update the fedora builds from released OVN branches.  We are close to ovn21.09, so I guess its fine.

https://bodhi.fedoraproject.org/updates/FEDORA-2021-60f76657b2

File: go-controller/pkg/ovn/ovn.go
Comment (hzhou8) [reply go-controller/pkg/ovn/ovn.go]: @dcbw I have decoupled the code from L3Gateway as suggested. Please take a look at the latest version.

**Comment by hzhou8** [comment]
@numansiddique thanks Numan. How am I supposed to use that fedora build?
@trozet @girishmg could you help triggering the workflow again? (sorry it is a little annoying having to ask maintainer to approve running workflows)

**Comment by girishmg** [comment]
@hzhou8 triggered the build.

I think the expectation is to update Dockerfile.Fedora with `ovn-21.06.0-14.fc33` .. obtained from the link that Numan shared.. when you are ready for final review..

**Comment by trozet** [comment]
@hzhou8 it is in koji: https://koji.fedoraproject.org/koji/packageinfo?packageID=28508

so just update the Dockerfile.fedora to point to the new build

**Comment by trozet** [comment]
@hzhou8 are there any upgrade concerns with this change? I noticed local gateway upgrade failed. It doesn't usually flake.

**Comment by hzhou8** [comment]
Thanks @girishmg @trozet . I updated the Dockerfile.fedora using the new ovn build just now.
For the upgrade tests, only the "local" one failed, but the "shared" one succeeded. I didn't figure out the link between the failure and my change. The logs doesn't tell much details. @girishmg triggered another run. So let's see.

I don't have any concerns w.r.t. upgrading, at least nothing should fail. However, I'd point out that due to the current limitation of ovn-controller incremental processing, it won't automatically remove local datapaths that are not needed anymore, so the scale benefit of this change won't be seen immediately after upgrading until recompute is triggered in ovn-controller (either manually or by some specific changes in DB).

**Comment by hzhou8** [comment]
Well, it fails again at the upgrade (local) tests, while the actual failure is a little different from the last run. To confirm if it is my change causing this failure, I created another PR that only contains the change of the ovn build version: https://github.com/ovn-org/ovn-kubernetes/pull/2440. @girishmg @trozet could you please trigger a CI run there?

**Comment by trozet** [comment]
@hzhou8 perhaps there is some incompatibility with local gateway mode still using DGP?

https://github.com/ovn-org/ovn-kubernetes/commit/ce2d312c7266c8d55173c9660002c67715a5444b#diff-78b4cf8349361d498f8f8440d8f11e2edb5d821c0d6f45ea7e053652b5c49b40R216

**Comment by hzhou8** [comment]
> @hzhou8 perhaps there is some incompatibility with local gateway mode still using DGP?
> 
> [ce2d312#diff-78b4cf8349361d498f8f8440d8f11e2edb5d821c0d6f45ea7e053652b5c49b40R216](https://github.com/ovn-org/ovn-kubernetes/commit/ce2d312c7266c8d55173c9660002c67715a5444b#diff-78b4cf8349361d498f8f8440d8f11e2edb5d821c0d6f45ea7e053652b5c49b40R216)

Thanks @trozet ! I think you are right. Here we still add DGP when gateway_mode is local:
https://github.com/ovn-org/ovn-kubernetes/commit/ce2d312c7266c8d55173c9660002c67715a5444b#diff-1f90f4648dc7d68c42854549c4257ddcfc9bd9a5b6fc480c74e3a23989914bc3R397

When multiple DGPs are used in LR, NAT is not supported yet. Even if we add general NAT support for multiple DGPs, DNAT_AND_SNAT (distributed NAT) also disables the optimization we can get from this change. So I will just add a check in this PR to avoid using DGPs when gateway_mode is local. We can remove the check if "local" mode is dropped in the future.

**Comment by trozet** [comment]
@hzhou8 would it be possible in the future to add the missing support so that local gw would also benefit from this change? We are looking to improve scale for older releases of OCP. From what I've heard, this fix seems like a massive improvement.

**Comment by hzhou8** [comment]
> @hzhou8 would it be possible in the future to add the missing support so that local gw would also benefit from this change? We are looking to improve scale for older releases of OCP. From what I've heard, this fix seems like a massive improvement.

@trozet Yes, it is possible, if the benefit is important for "local" mode, but it would need some more changes in OVN:
1) To make "local" mode continue to work, support NAT when multiple DGPs are used in OVN. This effort is ongoing AFAIK.
2) To avoid pulling unrelated datapaths to the node when distributed NAT is configured, add a new option that forces chassis-redirection so that user can set it for the DGPs that they know for sure won't need the distributed NAT processing. In ovn-k8s case it means all the DGPs connecting the node LSes. This change should be simple and I can add it later.

I hope it is ok to support "shared" mode first for now. I just pushed the change that makes sure gateway_mode is not local. Could you help trigger the workflow again?

**Comment by numansiddique** [comment]
Latest fedora build - https://bodhi.fedoraproject.org/updates/FEDORA-2021-4db76b17f4 - 
ovn-21.06.0-15.fc33

**Review by trozet — APPROVED** [review]
/lgtm

**Pull request merged**
